### ISSUE TITLE
Bad games being uploaded?

### ISSUE THREAD

**Issue by Hersmunch** [leela-zero/leela-zero#458]
Found a game uploaded that ends without passing or resignation. http://zero.sjeng.org/view/901a20869e379a162556b5965f498cef9f0c4575f2bc8391c563e9dd977fd03f?viewer=wgo
It is a short game that ends after a black move and reports a white win so perhaps two passes occurred but weren't put in the sgf?

````
(;GM[1]FF[4]RU[Chinese]DT[2017-12-21]SZ[19]KM[7.5]PB[Leela Zero 0.9 ../leela]PW[Human]RE[W+6.5]
--
  |  
  | ;B[jb];W[qq];B[iq];W[rd];B[nb];W[dd];B[mq];W[eq];B[ph];W[ge]
  | ;B[lf];W[db];B[pm];W[rp];B[ln];W[in];B[ig];W[qi];B[el];W[rb]
  | ;B[qp];W[oq];B[hf];W[oh];B[rq];W[pp];B[hp];W[fg];B[dg];W[pf]
  | ;B[ff];W[eg];B[go])
````

The player name doesn't show the hash... hmm
Wait I found another.
http://zero.sjeng.org/view/433aa775f5b4578aba79a571adf24aeea591143d289f89af6be6d2cfa9e149d0?viewer=wgo
and more:
http://zero.sjeng.org/view/377cfae402a169137ffecb3892ea086626692a8efa611d7b4d54023610139fa5?viewer=wgo
http://zero.sjeng.org/view/9543a673320074b86b2aa502560474801eed5dbb3d097cb7bc6fb629daa2f105?viewer=wgo

Is someone uploading bad games?

**Comment by kimitsu** [comment]
Definitely. 2421 games per hour with 216 clients seem like a higher-than-usual number, and you do hit these sgfs quite often (>50%) on the web page as last uploaded self-play, as of now. The training data will get spammed by those. Hope @gcp can filter that out...

**Comment by barrtgt** [comment]
according to https://docs.google.com/spreadsheets/d/1QhYWYQvXBhI5eHJm45fvf97CaV2rRmgrATwVgBFGKoU/htmlview#
we should expect about 1200 games per hour

**Comment by roy7** [comment]
Since the hash they are playing with is "../leela" I'd like to request a feature that the player name actually only show the first 8 characters of the hash itself, not the path to the weights file. Since "../leela" is useless to know what weights file they are using. :) @marcocalignano

**Comment by killerducky** [comment]
@roy7 I don't think any normal autogtp client will ever show a path like that. When autogtp runs leela it gives just the weight filename because its in the CWD.

**Comment by roy7** [comment]
Yeah I assume they must have manually edited the autogtp program to use a weights file from somewhere else, for whatever reason.

**Comment by barrtgt** [comment]
does autogtp check the hash of the weight file before it runs?

**Comment by roy7** [comment]
No but it doesn't matter if it did since if someone is intentionally interfering with the project they can just do whatever changes they want to autogtp anyway.

**Comment by barrtgt** [comment]
Time to blacklist IPs.

**Comment by roy7** [comment]
I'd like to believe it's just some sort of weird error state someone's client is in, but I realize that's unlikely. It's somewhat demoralizing to have people try to actively destroy an open source community project you volunteer a lot of time to.

**Comment by yufanyufan** [comment]
A few sugestion:
1. Have a encryption key embedded in pre-compiled binary, and sever only accept client from whitelisted key. This also ensure only latest version of client is used.

2. Only use single thread and included rng seed in the uploaded game, so every game is reproducilbe and verifiable. To make full usage of gpu, Autogtp can run multple game in parallel.

3. Sample 1% of game, and rerun them to test for attach or software bug.
.

**Comment by killerducky** [comment]
@yufanyufan for your 2nd suggestion see #426. autogtp already uses 1 thread.

Speaking of #426 and parsing weights, I am changing that now to be a little more robust. But I didn't plan on checking for weights having a path -- that is not supposed to happen and automatically fixing it would hide a problem. Like this one.

**Comment by killerducky** [comment]
BTW while testing the above I think I uploaded an sgf with "PB" twice instead of PB, PW. Oops.

**Comment by jjoshua2** [comment]
Instead of adding complicated DRM, what about just having people have people use a username/password to submit games? Than you can ban users and have trusted users, and when people switch IPs they will known to be the same entity with the same trust level. Sampling some random games from new players seems to be a good idea with how the games are selected and how many moves verified being not disclosed.

**Comment by gcp** [comment]
>A few suggestions:

1) You can just extract the key out of the binary. This is really easy.
2) This is implemented in the "next" branch.
3) "Rerun" is a bit problematic because the games aren't reproducible. See #300, #361.

**Comment by marcocalignano** [comment]
@roy7 I do not understand what you want. My games look like this:
```
(;GM[1]FF[4]RU[Chinese]DT[2017-12-18]SZ[19]KM[7.5]PB[Leela Zero 0.9 5166c094]PW[Leela Zero 0.9 0eb8b23b]RE[W+Resign]

;B[gl];W[pb];B[dp];W[ni];B[lq];W[rn];B[rm];W[rf];B[qn];W[bn]
;B[ro];W[hg];B[gg];W[fb];B[hh];W[gh];B[nh];W[qo];B[sn];W[pr]
;B[oi];W[oh];B[ph];W[og];B[pg];W[jo];B[of];W[ng];B[nf];W[mh]
;B[mg];W[lg];B[mf];W[rp];B[nh];W[lh];B[hq];W[lf];B[le];W[ke]
;B[kd];W[je];B[jd];W[ld];B[me];W[id];B[ie];W[he];B[if];W[ig]
;B[hf];W[gf];B[jf];W[kf];B[jg];W[fg];B[ih];W[gg];B[ic];W[hd]
;B[hc];W[jh];B[gd];W[kg];B[lc];W[qd];B[po];W[qp];B[pp];W[pq]
;B[qq];W[rq];B[qr];W[rr];B[qs];W[pn];B[rs];W[qm];B[rn];W[oq]
;B[oo];W[on];B[no];W[nn];B[nq];W[mo];B[np];W[mn];B[ql];W[ps]
;B[pm];W[ss];B[jq];W[cl];B[ej];W[cd];B[nj];W[ch];B[cj];W[nr]
;B[bh];W[cg];B[bg];W[bf];B[cf];W[df];B[ce];W[be];B[ci];W[de]
;B[mr];W[mp];B[mq];W[gc];B[om];W[km];B[gb];W[fc];B[nm];W[mm]
;B[ml];W[ll];B[nl];W[lk];B[dn];W[jc];B[md];W[jb];B[ib];W[bp]
;B[bq];W[cq];B[dq];W[cp];B[cr];W[br];B[dr];W[aq];B[mi];W[li]
;B[dl];W[cm];B[dm];W[ck];B[dk];W[bk];B[ja];W[kb];B[lb];W[ka]
;B[ia];W[bi];B[bj];W[ai];B[aj];W[ak];B[ah];W[nb];B[kc];W[rh]
;B[rg];W[qg];B[qh];W[qf];B[la];W[ri];B[il];W[qi];B[lj];W[kj]
;B[mj];W[kk];B[in];W[io];B[ho];W[jn];B[hn];W[ip];B[hp];W[iq]
;B[ir];W[jr];B[kr];W[hr];B[js];W[gr];B[ij];W[gq];B[cn];W[bm]
;B[bo];W[co];B[do];W[ao];B[gj];W[qb];B[rk];W[sg];B[fa];W[eb]
;B[dh];W[dg];B[ea];W[db];B[da];W[cb];B[ca];W[bb];B[ns];W[or]
;B[ga];W[eh];B[jm];W[jl];B[im];W[kn];B[ms];W[di];B[dj];W[ik]
;B[hk];W[jk];B[hj];W[mk];B[kp];W[nk];B[ok];W[sp];B[ji];W[ki]
;B[bs];W[ar];B[mb];W[nc];B[pf];W[pe];B[qr];W[qq];B[od];W[oc]
;B[pd];W[qe];B[pc];W[qc];B[oe];W[qs];B[gi];W[fi];B[fj];W[ei]
;B[pi];W[qj];B[cs];W[pj];B[oj];W[pk];B[ol];W[pl];B[qk];W[rj]
;B[sk];W[sj];B[jj];W[na];B[ba];W[ab];B[ko];W[ln];B[lo];W[jp]
;B[lp];W[op];B[gp];W[os];B[fd];W[ed];B[af];W[ae];B[ag];W[nd]
;B[mc];W[ne];B[jg];W[jf];B[as];W[ng];B[og];W[is];B[jr];W[ks]
;B[ls];W[so];B[sl];W[bi];B[ai];W[aa];B[hs];W[gs];B[is];W[jc]
;B[fq];W[jb];B[kb];W[jc];B[fs];W[fr];B[er];W[fe];B[ii];W[ge]
;B[ma];W[oa];B[gd];W[fd];B[hl];W[hb];B[ha];W[sc];B[sr];W[sq]
;B[eg];W[ef];B[jb];W[qa];B[ie];W[if];B[si];W[sh];B[sf];W[se]
;B[re];W[rd];B[rc];W[rb];B[cf];W[ce];B[an];W[am];B[en];W[tt]
;B[al];W[bl];B[ds];W[tt];B[ra];W[sa];B[gn];W[tt];B[ad];W[tt]
;B[bd];W[ac];B[bc];W[cc];B[ad];W[bd];B[dc];W[tt];B[dd];W[tt]
;B[gr];W[tt];B[hm];W[tt];B[hi];W[tt];B[ks];W[tt];B[hr];W[tt]
;B[ni];W[tt];B[qm];W[tt];B[fk];W[tt];B[gs];W[tt];B[jc];W[tt]
;B[fm];W[tt];B[el];W[tt];B[es];W[tt])

```

And like you see the player have the 8 first characters of the hash.

**Comment by gcp** [comment]
>what about just having people use a username/password to submit games like fishtest? Than you can ban users

Users can just create new accounts.

>Sampling some random games from new players seems to be a good idea with how the games are selected and how many moves verified being not disclosed.

The whole question is how to do this. Once you figure that out, you might as well ban IPs.

**Comment by gcp** [comment]
Tangentially related to this problem, it is not at all so obvious to me what data one would have to submit to make sure the learning actually fails. I mean, spamming out the valid data works, but that is also the easiest to stop. (Assuming no distributed attack, which will always succeed given our resources)

**Comment by marcocalignano** [comment]
My provider changes my IP twice a day. If a ban IP is inside my IP pool my games risk to get rejected even if I was trusted before.

**Comment by justinwong098** [comment]
Sorry for asking such a basic question, but if the same RNG seed is provided, why are games still not reproducible across different machines?

**Comment by kimitsu** [comment]
> Tangentially related to this problem, it is not at all so obvious to me what data one would have to submit to make sure the learning actually fails. I mean, spamming out the valid data works, but that is also the easiest to stop. (Assuming no distributed attack, which will always succeed given our resources)

As far as I understand the training data is not really verified against the sgfs as of now, so you can technically upload normal sgfs but bad training data, for example with 500 identical positions with certain bad P or V. Sure, it might not halt the process, but will definitely hurt.

> "Rerun" is a bit problematic because the games aren't reproducible. See #300, #361.

As you have said, the games are reproducible per hardware/software setup, so you could group clients based on their output automatically, and have them verify each other's results within each group. That does not solve the general problem, though, because bad actors can claim their own group, and unless one of them is busted, they can operate successfully.

Another way is to run statistical tests on the data, like the one I proposed earlier, to look for the games that are outstanding from the main population. Sure, the method is not bulletproff, and there's the risk of reacting on edge cases, but I think it would be a reasonably effective compromise.

**Comment by justinwong098** [comment]
>See #300, #361.

Unfortunately, I've already read those threads, and I don't see any concrete answer. Or is it that no one currently knows the answer?

Part of my misunderstanding is in how the RNG is used by Leela. Is it used to randomly search different parts of the tree? If so, why not directly store the numbers generated by the RNG, instead of just the seed, since storing just the seed doesn't work so great? Or is storing the all those numbers infeasible, or unpractical?

**Comment by yufanyufan** [comment]
Even a game can't not be perfectetly reproduced, It still can be verified if majority of the moves are verified.

**Comment by justinwong098** [comment]
@yufanyufan 
How do you propose that we verify only some moves, without reproducing the whole game?

**Comment by yufanyufan** [comment]
@justinwong098 
You can use LZ to predict next move provided all all previous move in the sgf file, and check how many move are same. In this way, the float computation error does not accumulte.
You may need to record the rng seed for each move though, because rng number may be gernerated different times.

**Comment by justinwong098** [comment]
@yufanyufan 
Instead of comparing the moves, I wonder if it would be better to compare the probabilities generated by the policy network and calculate how similar they are.

**Comment by gcp** [comment]
>Or is it that no one currently knows the answer?

This. Well, sortof. We know it's likely due to rounding differences in compiler and OpenCL implementation. It would be good if it were at least possible to reduce the compiler difference better.

>If so, why not directly store the numbers generated by the RNG, instead of just the seed, since storing just the seed doesn't work so great?

Storing the seed works fine, the problem has nothing to do with the RNG.

>so you could group clients based on their output automatically, and have them verify each other's results within each group

I'm a bit afraid that unless the compiler difference when building from source is reduced, there's too many groups.

>Another way is to run statistical tests on the data, like the one I proposed earlier, to look for the games that are outstanding from the main population.

I'm working along these avenues yes.

**Comment by gcp** [comment]
>Instead of comparing the moves, I wonder if it would be better to compare the probabilities generated by the policy network and calculate how similar they are.

The output in the training data is not the probabilities from the policy network.

**Comment by qysnn** [comment]
Maybe we can manually input the moves in these games into leela to check whether they are just accidentally ended without resign or pass, or they are deliberately generated to sabotage our training. Even if we cannot reproduce exact move, the moves chosen by valid clients would still on top of the list given all the previous moves.

**Comment by gcp** [comment]
I have something working now and it's revealing a few other things aside from that spammer.

There are clients that never play full games, but always have resignation enabled. There are also clients that seem to resign later than they should.

**Comment by jjoshua2** [comment]
I could see those resigning issues as a grey area were people disagreed
with the resign settings and edited those. It probably wouldn't
substantially effect anything.
Maybe the server can assign whether a game should have resignation enabled.

**Comment by roy7** [comment]
The /next client does do that.

**Comment by killerducky** [comment]
@gcp just a reminder for many days my client was running with resign disabled so I could do the resign analysis. Now that it seems stable I haven't done that since maybe 5 days ago.

**Comment by roy7** [comment]
@killerducky If you have a fixed IP address would you email it to @gcp so he can tell which user you are? Don't post it here, best to not make IP addresses public. :)

**Comment by killerducky** [comment]
I'm on vacation so I can't check it until like next week. If necessary you could probably find me by looking for sgf files with pull #426 style comments in them. Probably from a couple different IPs since I'm now running it from my parent's house.
C[Leela Zero options: -p 1600 -r 1 -m 30 -t 1 -d -n --noponder -s 8972309646146436749 -g -q -w 657f91c714d41ea1d0cf86ec5c2f1e69650cff3ee551dcf046b0062e4a285010]

**Comment by Azirine** [comment]
Someone has been uploading complied versions of the next branch on a Chinese forum, can you check if these are a problem? https://ufile.io/763ua

**Comment by gcp** [comment]
>just a reminder for many days my client was running with resign disabled so I could do the resign analysis.

I think this is the opposite of what I saw.

**Comment by gcp** [comment]
>I could see those resigning issues as a grey area were people disagreed with the resign settings and edited those.

The thing is that the the default resign threshold is "1" and an integer, so you can't actually lower it with just changing settings - you have to edit the program.

The games without resigning looked benign. So this is probably not a malicious user. But if I enable automated filtering, it's not going to be able to guess intent like I am.

**Comment by gcp** [comment]
>Someone has been uploading complied versions of the next branch on a Chinese forum, can you check if these are a problem?

Am I going to run a random binary on my system? No.

**Comment by OmnipotentEntity** [comment]
Could this be mitigated by requiring some sort of registration system to submit games?

**Comment by Marcin1960** [comment]
It could be an easy and unobtrusive "registration". When you run leelaz first time it ask you email, then you receive a key like 5 digits, then next time it asks you for the key. Both values are saved in a file in your directory and you are not bothered again as each time they are used to start. If you lose it another key will be send to you.

That way you can have a specific user accounts with IPs included. Several changing IP can be stored and they should be similar or belong to a few groups when using different locations.

**Comment by arnaudgo** [comment]
Hello,

IMHO, before taking any counter-measure, the leela client should perform consistency checks itself and make a decision according the well-known patterns: 1) flag the game bad and send it 2) discard the
And send metrics back to the server.

Arnaud

**Comment by marcocalignano** [comment]
The client is open source so the attacker can invalidates all the check it does.

**Comment by arnaudgo** [comment]
Of course @marcocalignano but still it's a convenient way to spot if it's due to human misbehaviours or due to software misbehaviours. It won't prevent much from the latter indeed but will help for the former.

**Comment by marcocalignano** [comment]
But how?

**Comment by arnaudgo** [comment]
Hello @marcocalignano 

Am I saying something wrong or stupid ?

If pattern A (A: a game ends without passing or resigning), it triggers a game flag "bad".
The client discards them or not. If it discards them then it doesn't alter learning.
If it sends them, the server can keep track of the bad games rate for that pattern.
And only an intentional spam would lead to games with pattern A without that flag since the attacker would modify the source to make pattern A still working for sabotage the learning. Or to find another way to corrupt the SGF data or whaever would interfere with the progress of LZ.

Arnaud

**Comment by roy7** [comment]
@marcocalignano The point is some people are running the client and submitting bad data, not because they are an attacker, but because their OpenCL driver doesn't work right and they have no idea their data is bad. They think they are helping out. The client needs to detect this sort of corrupt GPU state so it can filter the data out before it comes into the database.

**Comment by gcp** [comment]
>Could this be mitigated by requiring some sort of registration system to submit games?

As has been pointed out multiple times: no, it doesn't help at all.

**Comment by gcp** [comment]
>The client needs to detect this sort of corrupt GPU state so it can filter the data out before it comes into the database.

I wrote the code for this about a week ago and it is in the "next" client.  https://github.com/gcp/leela-zero/commit/133da087bd154bccf54b5531702b52627241b06d

Clients with broken OpenCL drivers will self-terminate.

**Comment by roy7** [comment]
Oh I didn't realize. Great. I guess the bad clients just haven't updated recently. In fact, any of those "movescount == null" results on the match games list are people who haven't updated /next client since @marcocalignano fixed that bug a week or more ago.

@gcp Should we increase client version required when I send out get-task commands to force everyone on /next to upgrade? It won't hurt release people who don't use /get-task yet.

**Comment by gcp** [comment]
>It won't hurt release people who don't use /get-task yet.

We could do that at some point but I didn't bump the client version on next when adding that code.

**Comment by marcocalignano** [comment]
But 
increasing the version of the client does not force them to recompile the leelaz engine. 
Would be better to increase the leelaz version to 0.10 then

**Comment by gcp** [comment]
>But increasing the version of the client does not force them to recompile the leelaz engine.

AutoGTP can request a minimum Leelaz version.

**Comment by marcocalignano** [comment]
@gcp Actually in the new code the leelaz engine version is taken from the json data.

**Comment by jkiliani** [comment]
The issue of bad games seems quite urgent at the moment. When I look through the currently submitted self-play games, there seems to be a clear split of what I expected from a network of roughly 1-dan strength, and a lot of games that look far weaker. At least some clients must be running a binary that produces weak search output, probably correlated with the massive increase of games submitted today (#557)

**Comment by marcocalignano** [comment]
I use the next branch from now Do I have to do something different?

**Comment by gcp** [comment]
2 games from the IPs submitting by far the most games (which are not-so-coincidentally being flagged as suspicious):

(;GM[1]FF[4]RU[Chinese]DT[2018-01-04]SZ[19]KM[7.5]PB[Leela Zero 0.9 55bf55dc]PW[Human]RE[W+Resign]\n\n;B[pq];W[ce];B[pd];W[dp];B[ii];W[ef];B[np];W[fc];B[hc];W[ji]\n;B[qf];W[lc];B[cq];W[dq];B[cp];W[co];B[bo];W[cn];B[bn];W[cm]\n;B[jj];W[bm];B[fh];W[dh];B[jc];W[cr];B[br];W[dr];B[nc];W[mp]\n;B[mo];W[lp];B[no];W[lo];B[iq];W[ll];B[ln];W[kn];B[km];W[lm]\n;B[mn];W[jn];B[kq];W[ij];B[jk];W[ik];B[ki];W[jh];B[il];W[jl]\n;B[hl];W[kj];B[kk];W[kl];B[lj];W[lk];B[kj];W[hj];B[mj];W[hh]\n;B[ho];W[hn];B[gn];W[in];B[gm];W[go];B[hp];W[fo];B[ko];W[nl]\n;B[gk];W[gj];B[fj];W[fk];B[ej];W[fl];B[em];W[gl];B[eo];W[fp]\n;B[kg];W[hf];B[hm];W[hk];B[el];W[kp];B[jo];W[jp];B[io];W[jq]\n;B[ir];W[mr];B[kr];W[jr];B[js];W[nq];B[oq];W[or];B[pr];W[lr]\n;B[ks];W[lq];B[os];W[nr];B[ek];W[jm];B[do];W[en];B[dn];W[fn]\n;B[dm];W[fm];B[cl];W[bl];B[bk];W[ck];B[cj];W[kh];B[lh];W[lg]\n;B[mh];W[kf];B[jg];W[ih];B[jf];W[bj];B[dk];W[ak];B[ck];W[je]\n;B[ke];W[kd];B[lf];W[jd];B[ls];W[bi];B[gq];W[ms];B[ie];W[is]\n;B[id];W[if];B[pl];W[le];B[mg];W[pc];B[oc];W[qd];B[qc];W[pb]\n;B[qb];W[pe];B[od];W[qe];B[rf];W[me];B[kc];W[ld];B[lb];W[mb]\n;B[kb];W[oe];B[ob];W[og];B[qi];W[oi];B[nf];W[ne];B[oj];W[ib]\n;B[hb];W[ic];B[hd];W[pi];B[pj];W[qh];B[rh];W[ri];B[qj];W[rg]\n;B[ph];W[sh];B[pg];W[oh];B[qg];W[rh];B[se];W[re];B[sd];W[rc]\n;B[sf];W[rb];B[pa];W[rd];B[of];W[pf];B[ng];W[sg];B[ni];W[sc]\n;B[mc];W[ep])

(;GM[1]FF[4]RU[Chinese]DT[2018-01-04]SZ[19]KM[7.5]PB[Leela Zero 0.9 55bf55dc]PW[Human]RE[W+Resign];B[ho];W[bm];B[pd];W[dd];B[pp];W[sf];B[dp];W[fc];B[cd];W[cc];B[bc];W[bd];B[og];W[ce];B[lq];W[ld];B[ng];W[ck];B[ql];W[qf];B[oq];W[qd];B[pe];W[pc];B[gh];W[oc];B[qh];W[mo];B[pm];W[nl];B[ll];W[on];B[pn];W[oo];B[ln];W[kp];B[lp];W[po];B[qo];W[lo];B[ko];W[kn];B[jo];W[qp];B[qq];W[rp];B[ro];W[lm];B[km];W[mn];B[jn];W[rq];B[qr];W[nq];B[nr];W[kl];B[ln];W[op];B[mm];W[pq];B[or];W[pr];B[ps];W[pp];B[nm];W[rr];B[mq];W[fk];B[qs];W[rs];B[os];W[np];B[ge];W[mr];B[lr];W[ns];B[dn];W[fm];B[cm];W[cl];B[bn];W[dm];B[cn];W[en];B[eo];W[dr];B[qc];W[qe];B[qb];W[pf];B[oe];W[mc];B[nd];W[nc];B[rd];W[fr];B[re];W[rf];B[rc];W[rh];B[qi];W[ri];B[rj];W[qj];B[pj];W[hd];B[he];W[id];B[ie];W[je];B[jf];W[kf];B[jg];W[cq];B[eh];W[ch];B[kg];W[lf];B[lg];W[pb];B[sb];W[hk];B[jd];W[ke];B[jc];W[gd];B[lb];W[mb];B[hb];W[gb];B[jb];W[lc];B[kb];W[mf];B[of];W[la];B[me];W[hq];B[ic];W[ka];B[mg];W[ma];B[fe];W[fd];B[gk];W[gl];B[gj];W[il];B[dl];W[em];B[dk];W[dj];B[cj];W[ek];B[di];W[el];B[ci];W[jq];B[fo];W[hj];B[gi];W[kk];B[lk];W[kj];B[hm];W[hl];B[gm];W[lj];B[mj];W[mi];B[nj];W[ak];B[ej];W[dk];B[bh];W[cg];B[bg];W[bf])

**Comment by gcp** [comment]
Last ~8400 games for 55bf55dc:

Good people: 271, bad people: 35
Good games: 5078, bad games: 3316

13% of clients, 35% of games are anomalous

**Comment by isty2e** [comment]
This is becoming serious... Did someone distribute bad binaries, presumably with closed PRs? If we are in a position to release another version, could it help?

**Comment by jkiliani** [comment]
If you have a way to quickly profile games, is there also bad data from earlier networks? I.e., does the long stall with 58da6176 have anything to do with data corruption?

**Comment by gcp** [comment]
I will release "next" soon, which adds self-checks. At this point, bumping the required version should bump off people who are using some unofficial client with bad settings.

**Comment by marcocalignano** [comment]
@gcp we need to change the URLs before you release. 
@roy7 can you give me the release server URLs for the json requests and the database uploads?

**Comment by gcp** [comment]
Searching for 58da61769a5fb12ff555ee927f16b50d4c24de73a1170a85df1b0d2d706bdef3
Good people: 808, bad people: 96
Good games: 53460, bad games: 57668

I'm a bit wary of filtering automatically, as nearly all of the top IPs are going to get all their games discarded. On the other hand, as you can see above, it does look like they are generating junk.

**Comment by gcp** [comment]
Another example:

Searching for 40b94cfed4f5387c94086a9d396ba2d35d03727863b5155c0181db6a00f7b1af
IP: x                           games:  1103 pv: [0.3648246371934275, 1.3284289999033598e-15, 0.51117342336129079, 5.1829065136464109e-10] tp: 4.2474985550000946e-21 Discarded
IP: y                          games:   536  pv: [0.23479521969472136, 7.6789536042583372e-18, 2.6452260666023549e-09, 0.11974634063964061] tp: 2.4655875805003956e-23 Discarded
IP: z                             games:   321 pv: [0.36064434287986469, 0.97813579204074319, 0.29147307936751976, 0.5986624080933185] tp: 0.6946427818926038 
IP: w                            games:   312 pv: [0.62352300773544789, 0.18589530398504675, 0.72402457501665918, 0.32675207915926441] tp: 0.5159759637402761 

"pv" are the p-values generated when seeing if the distribution of 4 (independent) move statistics could come from the same source distribution as the rest of the population. tp is the combined p-value from the 4 tests.

The top 2 IP's are highly significantly different.

**Comment by jkiliani** [comment]
If the problem was this bad but we could still generate 83015869 and 55bf55dc from the sum of this data, this indicates that the reinforcement learning algorithm must be pretty robust to bad data, at least while it is mixed with good data.

Is there a way to quantify _how_ bad the bad games are? I.e., is a bad game from the current network definitely weaker than a good game from a network 250k games ago?

**Comment by gcp** [comment]
Not sure how to quantity it, but it makes perfect sense to me that the learning is very robust to bad actors (and I said so a while ago).

If your junk is random, it will just get averaged out. If it is structured, the network will learn to recognize the bad data and won't spend much weights on it. Structuring the data so that it is guaranteed to "break" the learning may be unobvious and computationally intensive enough to be unattractive.

**Comment by ThorAvaTahr** [comment]
@gcp well, it should certainly slow down learning, gradients towards worse networks will happen, but they will not be selected. (right?)

**Comment by gcp** [comment]
>Or would that be a related to this issue?

I think that is unrelated. The best networks come from the same data. My best understanding is that the network can be on the edge regarding some move decision, and have them seem fairly equal in terms of training error, but they actually have a large effect on game play strength. (This is an argument to always use the latest network - it may feed back the data to make the network decision on this informed, rather then the current "wiggling" where we select it with a test when it happens to make the right decision)

**Comment by gcp** [comment]
>well, it should certainly slow down learning

Sure, it adds some noise. Also given that we plot strength over total games, obviously 35-50% of those being junk makes the graph seem to be a bit flatter than it really is.

**Comment by zjuwyz** [comment]
Maybe it's a good idea to show the last 100 or 200 recently uploaded self-playing games as we did on match games so that people can keep an eye on them.

**Comment by ThorAvaTahr** [comment]
> obviously 35-50% of those being junk makes the graph seem to be a bit flatter than it really is.

yeah, as you seem able to quantify (at least approximately) the number of bad games, it could be good to maintain an additional graph that corrects for this, that allows better comparison with the AGZ paper. Also I am not sure whether you already have a decision metric for changing the learning parameters (after 40 hours in the AGZ paper), but this could be relevant for that as well.

**Comment by thynson** [comment]
Implementation of #300 help us filtering bad guy. We could request the client to use given random seed to do self play, and then let the server assign same random seed to two or more different IPs, to verify if a game is junk.

**Comment by jkiliani** [comment]
@thynson Playing the same game twice seems excessive for verification even if feasible (which it currently isn't since same seed on different machines currently doesn't give repeatability). It would suffice to sample a few positions per game and verify that the probability maps produced in those are plausible.

**Comment by thynson** [comment]
@jkiliani That's another good way. But I don't know if the has enough computational resource to do so, and I just prefer to do it distributively.

**Comment by thynson** [comment]
@jkiliani It could be a more clever approach that we can add a new task to ask the client to download a partial sgf which was uploaded by another client before, and verify if the next move was same with the uploaded one.

**Comment by gcp** [comment]
Reproducible games would give very solid confirmation of which games are good and which ones are not. It would still be nice to have.

(Even if there are other ways to find bad clients that spam a lot of games)

**Comment by MaxMaki** [comment]
Could it be people running sethtroisis or NTT123s GPU optimizations, either running it incorrectly or some bug there?

Sethtroisi says in this reddit thread that he's getting 59 seconds per move (1400 games per day) on a GTX 1080 TI and that it's a 3x speedup roughly. That would fall in line with what you're saying that those clients are submitting 2-4x more?

https://www.reddit.com/r/cbaduk/comments/7n4j8j/1_day_1451_games_7kwh_050/

I'm sure some other people on that reddit have decided to try those improvements, and maybe they messed up something up?

**Comment by gcp** [comment]
Sure, quite possibly. Some people appear to be downloading random binaries from random places, so god knows what they're running.

**Comment by Tsun-hung** [comment]
there are still 2673 games uploaded in past hour, are some of them bad games?

**Comment by isty2e** [comment]
So the thing would be warn those people to stop using that specific client. Even when the next release is out they could be just changing the version numbers, so what about sending warning messages in the JSON based on the IP?

**Comment by killerducky** [comment]
We should wait for the release which auto detects the OpenCL driver bug. We should eliminate a known cause of bad games before thinking about other reasons.

**Comment by isty2e** [comment]
@killerducky That will help a lot, but look at this:

>The top 2 IP's are highly significantly different. They're also "somehow" generating 2-4x as much games as the next best client. The next clients look perfectly average.

I wonder if this is the driver issue after all.

**Comment by roy7** [comment]
@marcocalignano zero.sjeng.org uses almost identical code to zero-test.sjeng.org at all times. Any brand new changes I make I'll keep on zero-test for a few days at most in case something breaks, but then I push it over to live. You can use zero.sjeng.org for any/all purposes immediately, it's actually identical right now to test server.

**Comment by roy7** [comment]
@gcp Was any of the universities the one where the supecomputer is at that @alreadydone uses? From https://github.com/gcp/leela-zero/issues/505. Since he can pull down like 80 GPUs at once, if they had an OpenCL driver issue, that'd be a lot of bad games and he wouldn't have any way to know really. (I"m not sure if he's on /next with self check or not.)

**Comment by roy7** [comment]
@sethtroisi Might you email @gcp privately your IP address for the high volume game generation you do via Amazon or at home, so we can make sure there's not an issue with the games being submitted on our end?

**Comment by isty2e** [comment]
@roy7 Or what about an webpage where a user can enter his IP in and check if he is producing bad games?

**Comment by roy7** [comment]
Instead of a link to see your most recent game, maybe make it a link to see your most recent 100 self play games similar to how I output the VS links for matches?

**Comment by jkiliani** [comment]
@roy7: In the past, zero.sjeng.org offered the option to replay a self-play games created yourself, instead of the current option to view the most recent one by any client. If @gcp's plausibility check can be automated, could we not have a warning implemented on zero.sjeng.org if bad game data was detected coming from this IP?

**Comment by roy7** [comment]
@jkiliani It should still do this, but the link/text only appears if it found a game at your IP address.

**Comment by roy7** [comment]
But yes if we can detect bad IP addresses, we could make a list in the server to put up a big flashing red warning on the server to those IP addresses to try to contact someone who isn't actually malicious.

**Comment by marcocalignano** [comment]
@roy7 you should have got 8 request on the live server from me now and 'soon' you'll get some update. Can you please check?

**Comment by roy7** [comment]
@marcocalignano I'm sorry I'm at work and a bit behind, I can't mess with LZ as much as I'd like to today. I'll be free all evening of course (7 hours from now). I'll poke where I can as time permits today. But zero.sjeng.org and zero-test.sjeng.org are on identical code. Just listening to different port numbers.

**Comment by NightTusk** [comment]
Are these 'bad' clients also performing match games?

**Comment by roy7** [comment]
That'd let us know if they are using /next or not I guess.

**Comment by fffasttime** [comment]
@roy7 Maybe it's hard to contact them. Is there any method to let the client stop self play and show a warning when the client upload junk data? Then if someone changed the binary file and still upload bad data, we can sure he is malicious. It's also a best way to stop bad client waste their computing rescource.

**Comment by roy7** [comment]
If they were on the latest /next client I believe their SGF file would include things like full leelaz parameters/etc.

**Comment by odeint** [comment]
As @gcp can identify the affected (or malicious, as the case may be) clients by statistical methods, are their games being excluded from the training now? Overall I'm very impressed with the AlphaGo Zero algorithm, it really is very robust and scales quite well with the current, relatively small network. The anti-regression mechanisms really work.

**Comment by killerducky** [comment]
@gcp what is the false positive rate of your test? As measured by how many fails are detected from clients that are deemed good?

Also I get the impression your plan is to not release the test source code, to make it harder for attackers to defeat, is that right? I'd like to see the code just for curiosity but I understand if you don't want to release it.

**Comment by jkiliani** [comment]
I would definitely appreciate if a bad data warning was given automatically on zero.sjeng.org for the individual client once any games are submitted from the IP address. 

I just ran another self-play game after a considerable break since I was affected by the Mac OpenCL problem, now that self-testing and CPU support is on /next (I updated today, up to the caching commit). Unfortunately I'm unsure if even the new game is legitimate: http://zero.sjeng.org/view/9248d323a6704b62f8f09261ccc6cbe100eefa4f3250551a3dfaa3d050389f88?viewer=wgo

Looks suspicious to me as a Go player, since I regularly see much better games from other clients.

**Comment by MartinDevelopment** [comment]
@jkiliani That game looks legit to me.

**Comment by roy7** [comment]
@jkiliani I think move 14 is just from the randomness in first 30 moves. Overall it's playing about normal, and after move 30 play looks pretty rational all game long? Usually corrupted OpenCL driver gives seemingly random moves all game long.

**Comment by NelsonMinar** [comment]
Is the network actively blocking IPs now? My client stopped working about 10 hours ago (2018-01-03 20:31 Z), and I have no idea why. I just get the error `Could not talk to engine after launching.`  I've been running stock 0.9 with `./autogtp.exe -g 4` for a few weeks now on a pretty ordinary Windows 10 system with a GTX 1080.

I can promise I'm not intentionally submitting bad games or doing anything weird. Glad to work with someone to understand what the problem might be, maybe a bug in a standard Windows driver?

**Comment by roy7** [comment]
Could not talk to engine is a problem between autogtp and leelaz on your local system. That isn't a network/server related error that I know. Try running leelaz manually to see if it's still working?

**Comment by NelsonMinar** [comment]
Thanks for the quick replies and apologies for my bad guess. My leelaz is broken because Windows Defender is quarantining libopenblas.dll thinking it is the Win32/Peals.F!cl Trojan. Probably a false positive; it's doing that both with a fresh download and with the version I've been running for weeks. I'll investigate further and open a new bug report if appropriate.

**Comment by gcp** [comment]
>@gcp what is the false positive rate of your test? As measured by how many fails are detected from clients that are deemed good?

I don't see a good way of figuring this out. If I knew which clients were good, I wouldn't have to do analysis! 

I can set the p for the decision arbitrarily, and going from, for example, p=5% to 1% changes little (see the numbers I posted) because so many games are from a few clients that are so highly significant. If my maths are not wrong, this means any working client has a <=1% chance of being incorrectly flagged.

>Also I get the impression your plan is to not release the test source code, to make it harder for attackers to defeat, is that right?

I believe the method is sound in that even if an attacker saw it, he would have to generate a sufficient sample of "valid" games to know what kind of distribution the fake data should have. But as long as the client isn't reproducible, maybe I don't want to bet on that.

**Comment by gcp** [comment]
I'm not too keen of showing warnings on the webpage. The determination is done on the training machine, while it loads in the data to dump out the window. This means the information about the ip+network combos would need to get transferred back to the webserver. Given that the test is statistical, it *will* also cause false alerts. This isn't really an issue if the training data for a client is dropped for a network, but if we incorrectly show a warning that's going to just cause more confusion.

**Comment by killerducky** [comment]
You can assume for example, your own client is good, and do the test on games generated by your own client. Or pick an IP you trust. The theory is if p=1% it will give a 1% false positive. My question is what does it actually show? This is just a simple way to verify the statistical checker is working as intended.

Maybe this is overkill, from the results it sounds like it's probably doing what it should aleady. I understand there are many things interesting to do and limited time to do them all!

**Comment by sethtroisi** [comment]
@roy7 @gcp, I've tried really hard to verify that I'm producing the same results as OpenCL (I compute both and compare about half the time I compile and everytime I push) that being said some of the high level facts match my setup / code.

echo "my.id.add.res" | md5sum
6ca334862e64dd2f77e16084167bd17c

echo "sec.ond.com.put" | md5sum
fe370aff73eca66551669fd32d00ff9a

and I can often be fingerprinted as "XXX.244.YYY.ZZZ"

**Comment by alreadydone** [comment]
I've been uploading a large portion of the training games recently.
I started submitting games massively around Dec 30 at 1am GMT on the supercomputer Big Red II. If you see any clients suddenly began to upload many games in a short period of time, those are likely mine. (The supercomputer may not have a unique IP; is there an easy way to find out the IPs? I would prefer a command line option for `autogtp` to pass to the server a customized client ID.)
Because the jobs are submitted in batch, they can't be easily monitored and unfortunately I didn't save many sgf files. (I was still testing and didn't want to append `-k sgf` to every command.)
At the moment I was using the current master branch of gcp/leela-zero. There are 56 GPUs at the beginning which increase to 80GPUs in a few hours, and the speed is ~3.2 s/move per GPU (the GPUs only allow one `leelaz` to run). These are Tesla K20 GPUs with NVIDIA driver 352.68 and OpenCL support from CUDA (cudatoolkit/7.0.28-1.0502.10742.5.1). The `boost` version I used to compile was 1.55.0 (<1.58.0) and I had to do `$(CXX) -D_GLIBCXX_USE_CXX11_ABI=0` in Makefile to bypass some error. I don't know if these are concerns.

About one day later I switched to the `cache-nn` version of `leelaz` (but kept the `autogtp`, so I never submitted matches). At the moment I looked at a few games generated and they looked normal. The number of GPUs stayed roughly the same. Games are generated at ~2.5 s/move. About 1.5 days ago I merged `cache-nn` and `winograd` and have since been running this version at ~1.9 s/move.

About 2 days ago I found that more GPUs were being occupied and I could use less of them. I compiled a CPU-only version (left USE_OPENCL undefined) to run on 32-core CPU nodes on Big Red II. 32 instances of `autogtp` generate games at ~1 s/move per CPU node. About 8 hours ago I was utilizing ~125 CPU nodes (leading to the surge) but now only 85 nodes.

I hope the above information can help identify the problem. By the way, where can I see the training games?

**Comment by roy7** [comment]
@alreadydone Ok I found you. Man you are pumping in the games. One random one I looked at seemed fine, here is the one I looked at:

http://zero.sjeng.org/view/b60e0c339515c6afcaea2d5e1d59878145a1b62bbaf955ecce55189a64c0a2fa?viewer=wgo

You've uploaded 10537 games in the past 24 hours, and 1055 games in the past 1 hour.

**Comment by sethtroisi** [comment]
Last night I had very stable date suggesting 155 ms/move => 52 seconds/game => ~335 moves/game.
This also matches with the last resign analysis run I saw + the anecdotal game above.

1055 games/hour * 335 moves/game ~= 98 moves/second
10537 games/day * 335 moves/game ~= 40 moves/second

(1s/move)* 32 CPUs + (1.9s/move) * 80 GPUs = 75 moves/second

It could be that @alreadydone found 32 more CPUs to run on
or that there's some fencepost counting problems
or that some games are being cut short

**Comment by prusswan** [comment]
It might be a good idea to track the client metrics like OS, GPU/CPU model, driver version to better understand the "bad" hardware/software configurations. Once the bad clients have been identified, next step can be informing them on what they can do to rectify the problem on their end. 

Also, identifying clients by IP is imprecise, so a basic account system should be implemented.

**Comment by alreadydone** [comment]
@roy7 I am glad to help; it's a pity to left those resource unused. I also started to save sgf files for my games and they look good.
Can you identify me back to the time when 40b94cfe was the best network? I am possibly among the top two and I was using GPUs so there potentially were OpenCL problems. Am I a single IP address / client number or not?

**Comment by alreadydone** [comment]
@sethtroisi Maybe I wasn't clear but I am running `autogtp` on 105x32=3360 CPU cores right now.

**Comment by roy7** [comment]
@alreadydone GCP purges the web server database every few days so I don't have old data really, just the past day or so I think. (It's possible less than 24 hours even.) All of the data is on his training machines though. So he'd have to check.

It's possible the 10K in past 24 hours is lower than expected because there isn't even a full day's data in the web server currently. Past 1 hour should be accurate though.

**Comment by killerducky** [comment]
@alreadydone can you check if the GPU+OpenCL drivers on the supercomputer use has the bug? Getting a definitive answer would probably ease people's minds. There are a number of ways to test for it, you could download and compile the next branch and then run a few autogtps or just some standalone leelas. Or you could run the master branch standalone and run the "heatmap" command many times see if it ever returns garbage. Another option would be to just wait for the next release and try autogtp then.

**Comment by alreadydone** [comment]
@killerducky I did some tests and indeed find something unusual: using [`winograd`](https://github.com/gcp/leela-zero/pull/523) gives different heatmap than not using it! If `winograd` is buggy that would mean my games submitted today are buggy, but not the earlier games.
I got the [weights.txt](https://sjeng.org/zero/best_v1.txt.zip) an hour ago and I think it should be 55bf55dc. Below are my test results, with BRII standing for Big Red II:

BRII, [cach-nn + winograd](https://github.com/alreadydone/leela-zero) (rebased on /next branch), CPU/GPU output:
>     Leela: heatmap
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   1  39   0   0   0   0   0   0   0   0   0   0   0  45   1   0   0
>       0   0  47 152   0   0   0   0   0   0   0   0   0   0   0 127  48   0   0
>       0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0
>       0   0  41 174   0   0   0   0   0   0   0   0   0   0   0 154  43   0   0
>       0   0   0  45   1   0   0   0   0   0   0   0   0   0   0  43   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>     pass: 0
>     winrate: 0.502203


BRII, Current /master branch, GPU output (CPU version not implemented)
BRII, [cache-nn](https://github.com/gcp/leela-zero/pull/471#issuecomment-354623029) (rebased on /next branch), CPU/GPU output
My PC, Current /master branch, Windows binary, official release, GPU output
My PC, /next branch, Windows binary compiled a few days ago, GPU output
(If /next branch autogtp do self-checks then CPU output should be the same) 
>     Leela: heatmap
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0  39   1   0   0   0   0   0   0   0   0   0   0  39   0   0   0
>       0   0  49 173   0   0   0   0   0   0   0   0   0   0   0 121  45   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0
>       0   0  42 161   0   0   0   0   0   0   0   0   0   0   0 160  44   0   0
>       0   0   0  42   1   0   0   0   0   0   0   0   0   0   0  44   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
>     pass: 0
>     winrate: 0.504186

**Comment by bood** [comment]
@alreadydone weights.txt is not 55bf55dc, it is a network trained from human games, but still good for tests.

I can confirm that on my both machines, the heatmap matches the latter one you posted, suggesting first one is somehow broken.

**Comment by alreadydone** [comment]
@killerducky @Ttl @bood @Roy7 According to discussions at #523 the problem seems to pertain to larger (6x128) networks only, so I think my games today will not be problematic after all and I can continue to run the winograd+cach-nn version. (I killed ~75% of the running jobs and they need to be restarted now.)

**Comment by bood** [comment]
that is to say, we still dont know the source of these bad games?

2018年1月5日 下午12:50，"alreadydone" <notifications@github.com>写道：

**Comment by killerducky** [comment]
@alreadydone discussion is spread out so maybe I missed it. But did you run next branch with no changes on GPU to check for OpenCL issues? I see heatmaps for next+extra stuff and master, but not next only. Also to detect the bug you need to use a GPU and 
1) play a real game using next only OR
2) run heatmap many times (~100?) and make sure they are all identical. They won't be slightly off they will be totally wrong.

**Comment by isty2e** [comment]
@killerducky He said the results were identical regardless of CPU or GPU, with winograd+cache-nn. And interestingly, with the Winograd convolution, the spurious behaviours of Intel GPUs on mac are gone.

Hopefully @gcp can clear this situation up? If he can provide a list of IPs to @roy7, it will be easier to figure out whose machine/client is flawed.

**Comment by killerducky** [comment]
Identical results means identical heatmaps? But did he run the heatmap multiple times? Usually it takes awhile for the bug to pop up, it might work the first few times you run it.

**Comment by isty2e** [comment]
@killerducky Oh, that is a fair point. I have no clue about his situation, but for me, heatmaps remain exactly the same upon repeated calls, even for Intel GPUs. So I suppose the winograd convolution itself is not problematic (at least for 5x64).

**Comment by alreadydone** [comment]
@killerducky So the self-check is done throughout the game, not just on startup? And autogtp will terminate if once a disagreement between CPU and GPU evaluations is found? Which commit(s) is this? I compiled autogtp in /next branch and it is running now.

**Comment by isty2e** [comment]
@alreadydone It is [this commit](https://github.com/gcp/leela-zero/commit/133da087bd154bccf54b5531702b52627241b06d). When a game is being played, it sometimes compare the OpenCL results with BLAS ones with a pre-set probability. If an error is detected, leelaz itself is terminated.

**Comment by alreadydone** [comment]
@killerducky @isty2e @roy7 @gcp Each of the three versions of leelaz I used (current /master, cache-nn, and cache-nn+winograd) passed the self-check by finishing two games in AutoGTP v10.

**Comment by gcp** [comment]
The problem with the self-check in the winograd version is that it checks the output against the CPU version of winograd. The error in the heatmap you posted seems fairly large.

**Comment by isty2e** [comment]
By the way, what is happening to the already uploaded dubious games? Are they excluded from training?

**Comment by thynson** [comment]
Found another junk game there: http://zero.sjeng.org/view/27f122edf057e1185293c1dd9fbe4242767ca9c324da4a754273e989033ecf76?viewer=wgo

**Comment by Hersmunch** [comment]
What’s with white’s 126 and 20 in this match game? http://zero-test.sjeng.org/viewmatch/6a60a35d38cc2d6bef7b271334a537d56131bfb5d333d06cd38211d93096584c?viewer=wgo

**Comment by Dorus** [comment]
More weird moves in those games, like the second game, not only move 2 but also 13 and possibly 31 are off.

I also noticed the last network is really bad with self atari. Not 100% sure if this is related to bad uploads, but still something t watch out for. A few networks ago it had a very high win % vs gnu go, and now that win% is dropping down again, i checked some games it it loses a lot to self atari.

http://www.yss-aya.com/cgos/viewer.cgi?19x19/SGF/2018/01/05/347159.sgf for example 129 is bad and 249 seals the deal,

**Comment by gcp** [comment]
>By the way, what is happening to the already uploaded dubious games? Are they excluded from training?

We only filtered a few IPs that seemed to be intentionally flooding bad games. I am still a bit too concerned about false positives to filter automatically (and am not able to spend a lot of time on this to begin with).

One thing I found for example was that if one of the "very fast" clients submits games that are off, it will noticeably skew the /distributions, which will cause other "very fast" clients to get flagged (the tests get more sensitive the more games we have from an IP). My code has a reasonable ability to recover by systematically reducing the set of clients considered for calculating what is nromal until it finds a consistent one, but this can't be done exhaustively. And that means the "very fast" clients with a lot of games are proportionally more likely to get flagged out, which is not cool.

**Comment by seopsx** [comment]
I may have found some corrupted match games.

My attention was drawn to a client who uploaded 2 recent match games with the same move count and similiar scores (63.5 and 66.5). I checked the games with the losing side weights (68824bbc) and noticed that many of the actual game moves weren't even considered.

http://zero.sjeng.org/viewmatch/43cba0db9b6de6e26655aec74466d882ccf10f5927b8c3e5dd2c9306300c3a60?viewer=wgo

Bad moves: C2, P16 (big drop in winrate)

http://zero.sjeng.org/viewmatch/09d5c9a22754eee22ac1902a855722094bb6f8b7a4f8bc740af134adb5d6fccb?viewer=wgo

Bad moves: E4, G7, F6, R3, P17

**Comment by wonderingabout** [comment]
yes, all the fake games have in common of being very very short (1-2 minutes per 200 moves), with a lot of first line moves, see : #1705 

i think the user generating these should be blocked and noticed in some way if possible.

**Comment by seopsx** [comment]
> yes, all the fake games have in common of being very very short (1-2 minutes per 200 moves), with a lot of first line moves, see : #1705
> 
> i think the user generating these should be blocked and noticed in some way if possible.

Some of the fake match games I've seen have only couple of slack non-zero moves so they are not so easy to spot. And self-play is even harder because there is randomness involved.

We should be aware of this and try to monitor new games and hope that this perpetrator realizes to do something beneficial with the resources used in this nonsense.

**Comment by wonderingabout** [comment]
its possible that all this is just caused by a bug,

so trying to communicate with the client may be for the best

but at some point, regardless of this problem, leela zero project should increase its security game filtering from autogtp at some point

**Comment by seopsx** [comment]
Can somebody who has access to the training server confirm that this fake game issue https://github.com/gcp/leela-zero/issues/458#issuecomment-426521756 has been taken care of? @gcp @roy7

**Comment by hanamichigo** [comment]
Maybe we can train a network that will expose fake games? We can use it to immediately filter out on the server-side those obviously fake games such as shown on the images above, and as for those that are not-so-obvious, this network should still be able to flag suspicious games, for which additional tests by peers (or some other automated verification program) that will provide the answer on a high confidence level).

**Comment by Hersmunch** [comment]
Noticed this game from the latest match vs ELF https://zero.sjeng.org/viewmatch/9d166366fea7a4512b2309763916071039b8e0efc9a604098bb8752bc19ae8d9?viewer=wgo

**Comment by wonderingabout** [comment]
@Hersmunch 

wow thats a huge fish
nice catch !
